### ISSUE TITLE
unknown word and pos tags : ignore warning?

### ISSUE THREAD

**Issue by wadkar** [dsg-bielefeld/deep_disfluency#2]
This is more of a 'figuring out' how to use(tweak?) your 2015 model question.

#### some context:
My ultimate goal is to detect and remove disfluencies from a sample sentence. The sentences were gathered from crowdworkers' transcription of a conversation. Often times, crowdworkers try to faithfully transcribe a conversation and include disfluencies ("uh", "um", "so") in the transcription. I would like to remove these.

#### current setup:
I am using CoreNLP's API to segment the transcription into sentences, and get words and POS tags. I am following the first example in the `demo.py` and looping over each sentence that I get from CoreNLP and call `disf.tag_new_word` with pairs of word and POS tag. I call `disf.reset()` after each sentence.

#### sample code example:
Here's a made up example code describing my pipeline:
```python
crowd_text = """
this is transcription by uh, you know, crowdworkers.
they mostly do nice job but uh, well, not always.
it's because they tend to be inconstrained - i mean, they're inconsistent.
"""
snlp_client = CoreNLPClient(default_annotator=['ssplit', 'pos'])
anno_text = snlp_client.annotate(crowd_text)
for sent in anno_text.sentences:
    for word, tag in zip(sent.words, sent.tags):
        disf.tag_new_word(word=word, pos=tag, diff_only=True)
    for word, disf_tag in zip(sent.words, disf.output_tags):
        if '<rm' in disf_tag:
            print 'remove this word: %s because %s tag' % (word, disf_tag)
    disf.reset() # reset after each sentence
```

#### questions:
- I keep getting plenty of warnings such as `unknown word dude` and `unknown pos -LRB-`. I don't suppose I should ignore all of them?

- I see `corpus/BNC_swda_converter.py` converts BNC tags to PennTreebank tags. I thought CoreNLP yields PennTreebank tags and hence I don't need a converter. Obviously, I am missing something here - but I am not sure where to look. Does `swda` define its own version of PTB tags? Any help here would be great!

- Speaking of `corpus/BNC_swda_converter.py`; I suspect that I at least need to `compresses words together which were separated morphologically "don't", "isn't" etc.` (line 118). Am I thinking in the right direction here? Perhaps I could reuse that bit of your code?

- I thought `unknown word [X]` is telling me that the model does not have an embedding for word `[X]`. In which case, do I need to pre-train the word embeddings to account for new words found in my dataset?
On a related note, the first paragraph in your 2015 paper's section 3.1 "Architecture and parameters" says that pre-training with `'clean' language model [is] marginally more effective`. As I mentioned earlier, I was hoping to reuse your model as it is for now. Would you recommend that I look into pre-training for my use case? How do I go about it?

Thanks again for the very well documented code. I am learning quite a bit by simply reading through this repository. If you rather prefer email, I can be reached at wadkar@cmu.edu

Cheers,
-Sudhi

**Comment by julianhough** [comment]
Hi Sudhi,

Your usage looks good here! If you want to filter our the stand alone edit terms 'um's, 'uh's' and 'you know's, 'like' etc. which are not in a repair, you need to exclude anything tagged with `<e/>` too. Answers to your q's:

- When a word or POS tag is unknown/not used in training the model you'll get that warning, in that internally the tagger will just use the `<unk/>` token. This is not a problem as such as there are always a long tail of rare words in corpora where it doesn't make sense to have them internally as they're so sparse.

- Which PTB POS tags are not covered? You can see which POS tags are being used in `deep_disfluency\data\tag_representations` where the `pos_rep` files are the POS tag ones. The standard POS tag set is in `swbd_pos_rep.csv` there. If there are different formats, you have to normalize them to that format there, and if they're rarer tags, they may be unknown, as above.

- Yes, you would have to compress those kinds of morphemes together and remove punctuation. That's not done internally at the moment, only conversion to lower case. you can look at `deep_disfluency\data\tag_representations\swbd_word_rep.csv` to see which words are known in the right system.

- Yes, it doesn't have the embedding and hasn't seen the word in training according to a rare word criteria whereby if the word is not seen at least twice in the training data, we don't use it. As I say above, this isn't a big problem, as a it's designed to deal with unknown words. However, if you want a bigger vocab, you would have to train models (and embeddings) from scratch.

I'm happy to carry on the conversation on here for now. I really appreciate your input as to how to improve it.

Best,
Julian
